text/0148-sampling-probability.md

```diff
@@ -0,0 +1,209 @@
+
+### Example: Dapper tracing
+
+Google's [Dapper](https://research.google/pubs/pub36356/) tracing
+system describes the use of sampling to control the cost of trace
+collection at scale.
+
+### Example: Statsd metrics
+
+A Statsd counter event appears as a line of text, for example a metric
+named `name` is incremented by `increment` using a counter event (`c`)
+with the given `sample_rate`.
+
+```
+name:increment|c|@sample_rate
+```
+
+For example, a count of 100 that was selected for a 1-in-10 simple
+random sampling scheme will arrive as:
+
+```
+counter:100|c|@0.1
+```
+
+Probability 0.1 leads to a `sample_count` of 10.  Assuming the sample
+was selected using an unbiased algorithm (as by a "fair coin"), we can
+interpret this event as probabilistically equal to a count of `100/0.1
+= 1000`.
+
+## Internal details
+
+The statistical foundation of this technique is known as the
+Horvitz-Thompson estimator ("A Generalization of Sampling Without
+Replacement From a Finite Universe", JSTOR 1952).  The
+Horvitz-Thompson technique works with _unequal probability sampling_
+designs, enabling a variety of techniques for controlling properties
+of the sample.  
+
+For example, you can sample 100% of error events while sampling 1% of
+non-error events, and the interpretation of `sample_count` will be
+correct.
+
+### Bias, variance, and sampling errors
+
+There is a fundamental tradeoff between bias and variance in
+statistics.  The use of unbiased sampling leads unavoidably to
+increased variance.
+
+Estimating sampling errors and variance is out of scope for this
+proposal.  We are satisfied with the unbiased property, which guarantees
+that the expected value of totals derived from the sample equals the
+true population totals.  This means that statistics derived from
+sample data in this way are always accurate, and that more data will
+always improve precision.
+
+### Non-probabilistic rate-limiters
+
+Rate-limiting stages in a telemetry collection pipeline interfere with
+sampling schemes when they operate in non-probabilistic ways.  When
+implementing a non-probabilistic form of rate-limiting, processors
+MUST set the `sample_count` to zero.
+
+### No Sampler configured
+
+When no Sampler is in place and all telemetry events pass to the
+output, the `sample_count` field SHOULD be set to 1 to indicate
+perfect representivity, indicating that no sampling was performed.
+
+### Applicability for tracing
+
+The sampling approach covered here dictates how to select root spans
+in a probabilistic way.  When recording root spans, the `sample_count`
+field should be set as described above.  The adjusted `sample_count`
+of the root span applies the trace, meaning the trace should be
+considered as representative of `sample_count` traces in the
+population.
+
+When non-root spans are recorded because they are part of an ongoing
+trace, they are considered non-probabilistic exemplars.  Non-root
+spans should have `sample_count` set to zero.
```

Comment:
This sounds like it's saying that the cumulative statistics (e.g. throughput or error rate) cannot be inferred for a service that is not a root?

Reply:
@akehlenbeck Agree! I left too much of a gap in this document by talking about what we record in the span event as opposed to what a tracing system can do to sample and re-sample spans at various levels in a trace. 

I remember the Dapper details you describe--unfortunately it wasn't discussed in the Dapper paper. I heard that @bogdandrutu may recall removing the `inverse_probability` field from trace propagation in the present-day Dapper system. (I hear other means are now used.)

The conversation about how to create new roots at interior nodes of a trace is important (and was to Dapper) when a high-throughput service occasionally calls a low-throughput service. Uniform sampling of the high-throughput service rarely traces the low-throughput service; starting new roots makes low-throughput interior nodes visible. 

In some ways, you could use all "true" roots in your trace collection to analyze the behavior of a certain class of interior span. This would be a lot like analyzing a stratified sample. You could also compute an independent sample for traces rooted at the same class of interior span.  You could use both data sets to analyze the behavior of that class of interior span, as they support estimatation of the same quantities.

Dapper chose a less expennsive alternative to what I wrote above,which allows spans sampled as parts of rooted traces to be mixed with spans sampled as parts of "new-roots".  I believe what Dapper did and you describe is probabilistically correct, and I believe the proposed `sample_count` field can be used to carry the intended meaning if this approach is taken.  The proof would be to show how the sume of `sample_count` over a stream of spans is expected to equal the true span count.

I think a new OTEP to discuss propagating "inclusion probability" (or its inverse) for recording in sample spans is called for.

Reply:
@yurishkuro Reproducing the three possible cases in your example:
1. X and Y are sampled in 10% of all cases. The `sample_count` for Y would be 1.
2. X is not sampled while Y is sampled in 0.1 * 0.9 = 9% of all cases. The `sample_count` of Y would be 1 / (0.9 * 0.1).
3. X and Y are both not sampled in 81% of all cases. Y is not recorded.

The expectation of the sample count for Y would therefore be

10% * 1 + 9% * (1 / (0.9 * 0.1)) = 1.1

showing that there is something wrong. Unbiased sampling would give an expected sample count of 1.

This quite simple example already shows that it's not that easy to get subtree sampling right.

Reply:
> I heard that @bogdandrutu may recall removing the `inverse_probability` field from trace propagation in the present-day Dapper system. (I hear other means are now used.)

Is anyone able to share details about these other means?

Reply:
> So perhaps in case 1 the right thing to record is neither `p=0` like Josh wrote nor `p=1` like I wrote, but just `p=0.1 and Y isn't a root`, where the "Y isn't a root" part indicates that we're in case 1 and we shouldn't combine with other cases.

That makes a lot of sense to me and seems sufficient for tracing systems that don't support promoting child spans with missing parents. I'd also be curious if there's a way to combine this with case 2, but I don't have any ideas myself :-)

Reply:
This thread led me to remove a bunch of text about tracing specifics. I understand now that there are several ways to convey sampling information when collecting traces. I am now giving one example of how this can be used in a tracing system where only roots are sampled. I've also expanded the text about Dapper by a bit to explain this thread, somewhat.